### ISSUE TITLE
Clip negative values to 0, or throw error

### ISSUE THREAD

**Issue by 36000** [neuroneural/brainchop-cli#26]
Hello! I am integrating this into pyAFQ (https://github.com/tractometry/pyAFQ). If all works well, I want this to be our default tool for brain extraction and WM/GM/CSF extraction. 

I ran into a minor error that is easy enough to avoid, but could also be dealt with here. Having negative values in the T1 image can cause cryptic errors during normalization. I think these should be clipped in the conform function. I am not sure why, but some Qsiprep preprocessed T1 images happen to have small negative values.

Best,
    John Kruper

**Comment by sergeyplis** [comment]
Hi John,

Would not it be more flexible if we added an option with some intuitive name, say `--zero-negative` / `-zn` ? For example, CT scans naturally have negative values.

Thanks,
Sergey

**Comment by neurolabusc** [comment]
@36000 can you share a sample MRI image showing negative signal with my [institutional email](https://sc.edu/study/colleges_schools/artsandsciences/psychology/our_people/directory/rorden_chris.php)? By definition, MRI magnitude images [must be zero or greater](https://mriquestions.com/real-v-imaginary.html), and can not be negative. As an analogy, a car can drive in many directions (including reverse) but the current distance from the origin can never be negative. It might be worth understanding the stage that causes negative values - I would suspect a higher order (spline, sinc, etc) interpolation, where the lobes of the kernel can cause [ringing artifacts](https://nbviewer.org/urls/dl.dropbox.com/s/s0nw827nc4kcnaa/Aliasing.ipynb) around high contrast borders. Since several brainchop tools use niimath to conform data, and since niimath can chain operations, the obvious fix is simply `thr 0` before a conform. This should work for all the brainchop models designed for T1w input. However, one will need to think carefully about the `mindgrab` model as it is designed to handle different modalities, including CT calibrated in Hounsfield units where negative values are expected.

The Python script below uses sinc interpolation to provide concrete examples. Here are some command line instructions that apply the script to a [NIfTI image](https://github.com/rordenlab/NIfTIspace) illustrating three different possible corrections that should be self explanatory.

```bash
$ python downSample.py RAS.nii.gz halfRAS.nii.gz 
Input image: RAS.nii.gz
  shape: (64, 79, 67)
  min: 0.000000, max: 92.553886
Output image (pre-save):
  shape: (32, 40, 34)
  min: -17.896427, max: 114.619156
Resliced image saved to: halfRAS.nii.gz

$ niimath halfRAS.nii.gz -thr 0 -conform chalfRAS
$ niimath halfRAS.nii.gz -scale01 -mul 255 -conform chalfRAS -odt char
$ niimath halfRAS.nii.gz -thr 0  -scale01  -mul 255  -conform chalfRAS -odt char
```



```python
#!/usr/bin/env python3
"""
Reslice a NIfTI image to half the resolution using sinc interpolation.
"""

import sys
import math
import numpy as np
import nibabel as nib
from scipy.fft import fftn, ifftn, fftshift


def sinc_resample(img_data, new_shape):
    old_shape = img_data.shape

    f_img = fftn(img_data)
    f_img = fftshift(f_img)

    new_f = np.zeros(new_shape, dtype=np.complex64)
    min_shape = np.minimum(new_shape, old_shape)

    start_old = (np.array(old_shape) - min_shape) // 2
    start_new = (np.array(new_shape) - min_shape) // 2
    end_old = start_old + min_shape
    end_new = start_new + min_shape

    new_f[start_new[0]:end_new[0],
          start_new[1]:end_new[1],
          start_new[2]:end_new[2]] = \
        f_img[start_old[0]:end_old[0],
              start_old[1]:end_old[1],
              start_old[2]:end_old[2]]

    new_f = fftshift(new_f)

    resampled = np.real(ifftn(new_f))

    # Scale intensities to preserve mean value
    scale_factor = np.prod(new_shape) / np.prod(old_shape)
    resampled *= scale_factor

    return resampled


    # Inverse FFT
    resampled = np.real(ifftn(new_f))

    return resampled


def main():
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} input.nii.gz output.nii.gz")
        sys.exit(1)

    in_file, out_file = sys.argv[1], sys.argv[2]

    # Load NIfTI image (apply scaling so values match FSL)
    img = nib.load(in_file)
    data = img.get_fdata(dtype=np.float32)
    affine = img.affine

    print(f"Input image: {in_file}")
    print(f"  shape: {data.shape}")
    print(f"  min: {data.min():.6f}, max: {data.max():.6f}")

    # Compute target shape (half size, ceil for odd dimensions)
    new_shape = tuple([math.ceil(s / 2) for s in data.shape])
    new_header = img.header.copy()
    new_header['scl_slope'] = 1
    new_header['scl_inter'] = 0

    # Perform sinc interpolation
    resampled_data = sinc_resample(data, new_shape)

    print(f"Output image (pre-save):")
    print(f"  shape: {resampled_data.shape}")
    print(f"  min: {resampled_data.min():.6f}, max: {resampled_data.max():.6f}")

    # Adjust affine to reflect new voxel size
    zoom_factors = np.array(img.shape) / np.array(new_shape)
    new_affine = affine.copy()
    new_affine[:3, :3] = affine[:3, :3] @ np.diag(zoom_factors)

    # Save new image
    nib.save(nib.Nifti1Image(resampled_data.astype(np.float32), new_affine, new_header), out_file)
    print(f"Resliced image saved to: {out_file}")


if __name__ == "__main__":
    main()

```

**Comment by 36000** [comment]
Hello Sergey and Chris,
    Thank you for the prompt replies. Let me show you the steps to reproduce the bug. I now think it is a problem with the Nifti Image and not brainchop, so you can close this issue, but perhaps you have some insight. The T1 image I was having trouble with is available through DIPY:
```
from dipy.data import fetch_hbn
import os.path as op

data_path = fetch_hbn(["NDARAA948VFH"])[1]
t1_path = op.join(data_path, "derivatives", "qsiprep", "sub-NDARAA948VFH", "anat", "sub-NDARAA948VFH_desc-preproc_T1w.nii.gz")
print(t1_path)
```
Here is a command to trigger the bug:
```
brainchop sub-NDARAA948VFH_desc-preproc_T1w.nii.gz -i -o masked_t1 -a mask -m mindgrab
```

Here is the warning:
```
/Users/john/miniconda3/envs/afq11/lib/python3.11/site-packages/brainchop/tiny_meshnet.py:22: RuntimeWarning: invalid value encountered in divide
  img = (img - qlow) / (qhigh - qlow)
```

And if you open mask.nii.gz, it is all 2's. From here, I tried manipulating the image to make it work, and immediately found that clipping to have all values above 0 fixed it. However, I just tried again, and found that simply loading the image and saving it again with nibabel fixes the problem (no need to clip). So maybe there is something wrong with how QSIprep saved the image.

**Comment by neurolabusc** [comment]
@36000 the upcoming release of niimath will address two edge cases illustrated by your demo. Thanks for sharing, this will make niimath more robust:

The T1w magnitude image has a tiny number (0.2839%) of extremely negative values:
```bash
% fslstats sub-NDARAA948VFH_desc-preproc_T1w.nii.gz -R
-104.038704 643.140808 
```
while the freesurfer conform function has some tolerance for extremely positive values (f_high), it does not clamp the most [negative values (f_low)](https://github.com/Deep-MI/FastSurfer/blob/8dc91377abea35260a98c3168522f22a0db4ccdb/FastSurferCNN/data_loader/conform.py#L476).

Modern tools give precedence to the 12-DoF S-Form over the 9-DoF Q-Form [spatial transform](https://gru.stanford.edu/doku.php/mrtools/coordinatetransforms). However, my updated niimath will fall back to the Q-Form if it valid and has a higher code or if the S-Form is bogus.

```bash
$ fslhd sub-NDARAA948VFH_desc-preproc_T1w.nii.gz 
...
qform_name	Scanner Anat
qform_code	1
qto_xyz:1	-1.000000 0.000000 0.000000 96.000000 
qto_xyz:2	0.000000 -1.000000 0.000000 96.000000 
qto_xyz:3	0.000000 0.000000 1.000000 -78.000000 
qto_xyz:4	0.000000 0.000000 0.000000 1.000000 
qform_xorient	Right-to-Left
qform_yorient	Anterior-to-Posterior
qform_zorient	Inferior-to-Superior
sform_name	Scanner Anat
sform_code	0
sto_xyz:1	0.000000 0.000000 0.000000 0.000000 
sto_xyz:2	0.000000 0.000000 0.000000 0.000000 
sto_xyz:3	0.000000 0.000000 0.000000 0.000000 
sto_xyz:4	0.000000 0.000000 0.000000 1.000000 
```

**Comment by 36000** [comment]
Thanks for the help! For now I will continue to clip to 0 on my end, good to see it will not be necessary in general. I have another question: are there plans for adding more models here in the future? what is the process for adding a model? I am thinking of better segmentations of the cerebellum and/or brainstem in particular (often edge cases, literally, in neuroimaging).

**Comment by sergeyplis** [comment]
@36000 John, training a model is a bit trickier. Not every model is able to run at its fullest within the constrained environments and the simplest way to get a model trained to your segmentation task is to follow our process of training our parameter efficient architecture.

Here is an old and obsolete guide on [how to train this architecture](https://medium.com/pytorch/catalyst-neuro-a-3d-brain-segmentation-pipeline-for-mri-b1bb1109276a) 

Since it was written we have discovered a much better way of training models built on this architecture. See if you can figure out how to use [this example](https://github.com/neuroneural/catalyst_example). The documentation is minimal but I can answer your questions. Frankly, the only tricky part is that we are training from `mongobd`, which considerably speeds up our experiments. However, you can replace the dataset and dataloader with your own classes. I highly recommend using `meshnet.py` and `meshnetAE.json` from the more modern example. It enables full volume training in memory efficient way and provides a more performant architecture in the json file.

You can get more information on how this architecture works and why it is parameter efficient in the following papers:

1. [End-to-end learning of brain tissue segmentation from imperfect labeling
](https://ieeexplore.ieee.org/abstract/document/7966333)
2. [State-of-the-Art Stroke Lesion Segmentation at 1/1000th of Parameters
](https://ieeexplore.ieee.org/abstract/document/10981054)
3. [MindGrab for BrainChop: Fast and Accurate Skull Stripping for Command Line and Browser
](https://arxiv.org/abs/2506.11860)
